js-apps/meep-frontend/src/js/containers/cfg/cfg-page-container.js

```diff
@@ -141,6 +148,141 @@ class CfgPageContainer extends Component {
     this.props.cfgElemClear();
   }
 
+  // CLONE
+  onCloneElement() {
+    this.props.cfgElemClone();
+  }
+
+  // CLONE ELEMENT, return new element name
+  cloneElement(element, newParentName, isRoot) {
+    let newElement = deepCopy(element);
+
+    var name = getElemFieldVal(element, FIELD_NAME);
+    if (isRoot === false) {
+      name = createUniqueName(this.props.cfg.table.entries, name + '-copy');
+      setElemFieldVal(newElement, FIELD_NAME, name);
+    }
+    setElemFieldVal(newElement, FIELD_PARENT, newParentName);
+
+    // add new element to scenario
+    // new id and label will be created as part of the addNewElementToScenario called by newScenarioElem
+    this.props.newScenarioElem(newElement, false);
+    return name;
+  }
+
+  // CLONE
+  onApplyCloneElement(element) {
```

Comment:
All scenario manipulations should be in the scenario-utils.js file. Please move this function and make sure the scenario state is updated in the correct order; i.e. update the scenario and set the updated state when completed successfully. This scenario state update should then trigger a refresh of the visualization.

Reply:
moved to proper location